state/protocol/badger/validity.go

```diff
@@ -199,29 +200,44 @@ func isValidEpochCommit(commit *flow.EpochCommit, setup *flow.EpochSetup) error
 	return nil
 }
 
+// isValidVersionBeacon validates internal structure of flow.VersionBeacon and in a context of given block,
+// which is assumed to be a block flow.VersionBeacon has been put into effect.
+// InvalidServiceEventError with appropriate message is returned if any validation fails.
+// Please refer to documentation of flow.VersionBeacon for a description of what's expected of a valid VersionBeacon.
 func isValidVersionBeacon(vb *flow.VersionBeacon, header *flow.Header) error {
 	if len(vb.RequiredVersions) == 0 {
 		return fmt.Errorf("required versions empty")
 	}
 
-	// first entry in a table must be a current version, so the height must be below the current block
-	if vb.RequiredVersions[0].Height > header.Height {
-		return fmt.Errorf("lowest required version height %d below current block's height %d", vb.RequiredVersions[0].Height, header.Height)
-	}
+	minimumThresholdHeight := header.Height + flow.VersionThreshold
 
 	// handle case when only one version is present
 	if len(vb.RequiredVersions) == 1 {
+		if vb.RequiredVersions[0].Height < minimumThresholdHeight {
```

Comment:
I'm wondering if we really want to enforce this constraint here. 

The `VersionThreshold` exists because we might occasionally not seal a block for an extended period of time. We assume that we will never have a sealing outage larger than 1000 blocks. So we make sure that the height distance between a new `VersionRequirement`'s addition to the table and its effective height is at least 1000. That way we guarantee that a `VersionRequirement` will be known to an EN before that EN executes the first height where the requirement is effective.

In the diagram below, we want to make sure that `C.height > B.height`, because the distance between `A` and `B` can vary due to sealing lags.
```
/- Height where VersionBeacon event is emitted
|           /- Height where VersionBeacon event is sealed, and known to EN
|           |           /- First height where VersionRequirement is effective
v           v           v 
A <- ... <- B <- ... <- C
```

However, when validating the service event, we are are comparing `B.height` to `C.height`. If a VersionBeacon is emitted with a "version requirement buffer" near the threshold, then this check might fail unpredictably, depending on the sealing lag.

For example, suppose that at height 987 (block A) we emit a VersionBeacon with `height=2000, version=v2`. We have included a buffer of 1,013 blocks so we satisfy the 1000-block constraint. But then there is a sealing lag lasting 50 blocks, so we don't seal the event until height 1037. Then, we will fail validation and throw an error here. **⚠️ With the current error handling, that error would halt the chain and cause an outage.**

Reply:
Thanks for this comment! 
There are two aspects here.
1. This validation happens not when finalizing, but when we extends the chain - so using your diagram, we compare A and C (B is used to index an event and we don't enforce threshold there). 
2. Error handling - or rather invalid input data handling. I assumed that we don't really expect invalid service events, hence halting a chain seemed like a good idea. However, now I'm not certain, but can't really think of an alternative. Shall we simply skip an event, risking possibly running invalid version?

Reply:
> This validation happens not when finalizing, but when we extends the chain

Agreed - the validation happens when extending the chain by inserting a block. However, we are inserting the block where the service event is processed (block B), not the the block where the service event was emitted (block A). 

When we call `isValidVersionBeacon` here:
https://github.com/onflow/flow-go/blob/1c05cb7f754267e6f5622fcf4378c182bf53abd4/state/protocol/badger/mutator.go#L896
`block` is the child of the block sealing the service event (equivalent to block B in my example diagram above). So in the validation, we are using `B.height` in our comparison.